elastic/beats
# Add `container_v2` input type as Beta

This new container V2 input is based on filestream and supports all the filestream configuration options plus the options the previous container supported.



## Checklist



~~- [ ] I have made corresponding change to the default configuration files~~
~~- [ ] I have added tests that prove my fix is effective or that my feature works~~

## How to test this PR locally



Run Filebeat with the following configuration:

```yaml
filebeat.inputs:
  - type: container
    paths:
      - "/logs/*"
  - type: container_v2
    paths:
      - "/logs/*"
path.data: "/data"
logging:
  level: debug
output.console:
  enabled: true
```

Note: replace `/logs` with your directory that contains samples of Docker container logs. Replace `/data` with the folder where you're planning to store the state/registry.

You should have 2 sets of events like these:

container input

```json
{
  "@timestamp": "2023-05-12T07:46:33.603Z",
  "@metadata": {
    "beat": "filebeat",
    "type": "_doc",
    "version": "8.11.0"
  },
  "stream": "stdout",
  "message": "{\"thread_name\":\"redacted\",\"level\":\"INFO\",\"logger_name\":\"redacted\",\"message\":\"redacted\",\"end_of_batch\":true,\"unix_time\":1683877593602,\"logger_fqcn\":\"org.apache.logging.slf4j.Log4jLogger\",\"trace_id\":\"redacted\",\"utc_time\":\"2023-05-12 07:46:33.602 GMT\",\"vin\":\"redacted\",\"device_id\":\"redacted\",\"log_type\":\"webapp\"}",
  "log": {
    "offset": 0,
    "file": {
      "path": "redacted"
    }
  },
  "input": {
    "type": "container"
  },
  "ecs": {
    "version": "8.0.0"
  },
  "host": {
    "name": "redacted"
  },
  "agent": {
    "ephemeral_id": "22581cc5-a340-4194-81bc-6b4c73dec347",
    "id": "386dcd8a-477f-4d78-b1f6-2c0d362c39c2",
    "name": "redacted",
    "type": "filebeat",
    "version": "8.11.0"
  }
}
```

container V2 input

```json
{
  "@timestamp": "2023-05-12T07:46:33.603Z",
  "@metadata": {
    "beat": "filebeat",
    "type": "_doc",
    "version": "8.11.0"
  },
  "stream": "stdout",
  "message": "{\"thread_name\":\"redacted\",\"level\":\"INFO\",\"logger_name\":\"redacted\",\"message\":\"redacted\",\"end_of_batch\":true,\"unix_time\":1683877593602,\"logger_fqcn\":\"org.apache.logging.slf4j.Log4jLogger\",\"trace_id\":\"redacted\",\"utc_time\":\"2023-05-12 07:46:33.602 GMT\",\"vin\":\"redacted\",\"device_id\":\"redacted\",\"log_type\":\"webapp\"}",
  "log": {
    "offset": 0,
    "file": {
      "path": "redacted",
      "device_id": 16777234,
      "inode": 130554211,
      "fingerprint": "057224ae90e98a81df1d11b6e908edb8103aff95a2732cd0c68a58e8123c1dc3"
    }
  },
  "input": {
    "type": "container_v2"
  },
  "ecs": {
    "version": "8.0.0"
  },
  "host": {
    "name": "redacted"
  },
  "agent": {
    "ephemeral_id": "22581cc5-a340-4194-81bc-6b4c73dec347",
    "id": "386dcd8a-477f-4d78-b1f6-2c0d362c39c2",
    "name": "redacted",
    "type": "filebeat",
    "version": "8.11.0"
  }
}
```

with the only difference:

```diff
13c13,16
<       "path": "redacted"
---
>       "path": "redacted",
>       "device_id": 16777234,
>       "inode": 130554211,
>       "fingerprint": "057224ae90e98a81df1d11b6e908edb8103aff95a2732cd0c68a58e8123c1dc3"
17c20
<     "type": "container"
---
>     "type": "container_v2"
```

## Documentation

<img width="961" alt="Screenshot 2023-10-02 at 16 51 04" src="https://github.com/elastic/beats/assets/887952/90e7bd46-3945-4602-80f8-b2c888b795e0">
<img width="957" alt="Screenshot 2023-10-02 at 16 51 17" src="https://github.com/elastic/beats/assets/887952/4c29dcf5-2c2e-4637-b593-68263e8b1803">

## Related issues


- Relates #34393
